flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkLogicalRelFactories.scala

```diff
@@ -0,0 +1,288 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.calcite
+
+import org.apache.flink.table.calcite.FlinkRelFactories.{ExpandFactory, RankFactory, SinkFactory}
+import org.apache.flink.table.plan.nodes.calcite.RankRange
+import org.apache.flink.table.plan.nodes.calcite.RankType.RankType
+import org.apache.flink.table.plan.nodes.logical._
+import org.apache.flink.table.plan.schema.FlinkRelOptTable
+import org.apache.flink.table.sinks.TableSink
+
+import com.google.common.collect.ImmutableList
+import org.apache.calcite.plan.{Contexts, RelOptCluster, RelOptTable, RelTraitSet}
+import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeField}
+import org.apache.calcite.rel.core.RelFactories._
+import org.apache.calcite.rel.core._
+import org.apache.calcite.rel.logical._
+import org.apache.calcite.rel.{RelCollation, RelNode}
+import org.apache.calcite.rex._
+import org.apache.calcite.sql.SqlKind.{EXCEPT, INTERSECT, UNION}
+import org.apache.calcite.sql.{SemiJoinType, SqlKind}
+import org.apache.calcite.tools.{RelBuilder, RelBuilderFactory}
+import org.apache.calcite.util.ImmutableBitSet
+
+import java.util
+
+import scala.collection.JavaConversions._
+
+/**
+  * Contains factory interface and default implementation for creating various
+  * flink logical rel nodes.
+  */
+object FlinkLogicalRelFactories {
```

Comment:
do we need this?

Reply:
This class is used to create `FlinkLogicalRel` in rules. Currently, it only is used in `SplitAggregateRule` which is applied on `FlinkLogicalRel`. 
 If we do not provide this class, we need create each `FlinkLogicalRel` directly, instead of using methods in `RelBuilder`, like `project()`, `aggregate()`.

However, What I worry about is `FlinkLogicalRelFactories` can not create  a `FlinkLogicalRel` with new attribute field,  e.g. `FlinkLogicalAggregate` add `partialFinalType` to indicate whether to skip `SplitAggregateRule`. (How to remove `partialFinalType` from `FlinkLogicalAggregate` is another question.)